// mindee/src/product/us/bankCheck/bankCheckV1Page.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BankCheckV1Page = void 0;
const common_1 = require("../../../parsing/common");
const standard_1 = require("../../../parsing/standard");
const bankCheckV1Document_1 = require("./bankCheckV1Document");
/**
 * Bank Check API version 1.1 page data.
 */
class BankCheckV1Page extends bankCheckV1Document_1.BankCheckV1Document {
    constructor(rawPrediction, pageId) {
        super(rawPrediction, pageId);
        /** List of signature positions */
        this.signaturesPositions = [];
        this.checkPosition = new standard_1.PositionField({
            prediction: rawPrediction["check_position"],
            pageId: pageId,
        });
        rawPrediction["signatures_positions"].map((itemPrediction) => this.signaturesPositions.push(new standard_1.PositionField({
            prediction: itemPrediction,
            pageId: pageId,
        })));
    }
    toString() {
        const signaturesPositions = this.signaturesPositions.join("\n                      ");
        let outStr = `:Check Position: ${this.checkPosition}
:Signature Positions: ${signaturesPositions}`.trimEnd();
        outStr += "\n" + super.toString();
        return (0, common_1.cleanOutString)(outStr);
    }
}
exports.BankCheckV1Page = BankCheckV1Page;
