// mindee/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.imageOperations = exports.internal = exports.LocalResponse = exports.PageOptionsOperation = exports.UrlInput = exports.StreamInput = exports.PathInput = exports.BytesInput = exports.BufferInput = exports.Base64Input = exports.InputSource = exports.RagMetadata = exports.RawText = exports.JobResponse = exports.InferenceResponse = exports.InferenceFile = exports.Page = exports.Document = exports.Prediction = exports.Inference = exports.PredictResponse = exports.AsyncPredictResponse = exports.ClientV2 = exports.Client = exports.product = void 0;
exports.product = __importStar(require("./product"));
var client_1 = require("./client");
Object.defineProperty(exports, "Client", { enumerable: true, get: function () { return client_1.Client; } });
var clientV2_1 = require("./clientV2");
Object.defineProperty(exports, "ClientV2", { enumerable: true, get: function () { return clientV2_1.ClientV2; } });
var common_1 = require("./parsing/common");
Object.defineProperty(exports, "AsyncPredictResponse", { enumerable: true, get: function () { return common_1.AsyncPredictResponse; } });
Object.defineProperty(exports, "PredictResponse", { enumerable: true, get: function () { return common_1.PredictResponse; } });
Object.defineProperty(exports, "Inference", { enumerable: true, get: function () { return common_1.Inference; } });
Object.defineProperty(exports, "Prediction", { enumerable: true, get: function () { return common_1.Prediction; } });
Object.defineProperty(exports, "Document", { enumerable: true, get: function () { return common_1.Document; } });
Object.defineProperty(exports, "Page", { enumerable: true, get: function () { return common_1.Page; } });
var v2_1 = require("./parsing/v2");
Object.defineProperty(exports, "InferenceFile", { enumerable: true, get: function () { return v2_1.InferenceFile; } });
Object.defineProperty(exports, "InferenceResponse", { enumerable: true, get: function () { return v2_1.InferenceResponse; } });
Object.defineProperty(exports, "JobResponse", { enumerable: true, get: function () { return v2_1.JobResponse; } });
Object.defineProperty(exports, "RawText", { enumerable: true, get: function () { return v2_1.RawText; } });
Object.defineProperty(exports, "RagMetadata", { enumerable: true, get: function () { return v2_1.RagMetadata; } });
var input_1 = require("./input");
Object.defineProperty(exports, "InputSource", { enumerable: true, get: function () { return input_1.InputSource; } });
Object.defineProperty(exports, "Base64Input", { enumerable: true, get: function () { return input_1.Base64Input; } });
Object.defineProperty(exports, "BufferInput", { enumerable: true, get: function () { return input_1.BufferInput; } });
Object.defineProperty(exports, "BytesInput", { enumerable: true, get: function () { return input_1.BytesInput; } });
Object.defineProperty(exports, "PathInput", { enumerable: true, get: function () { return input_1.PathInput; } });
Object.defineProperty(exports, "StreamInput", { enumerable: true, get: function () { return input_1.StreamInput; } });
Object.defineProperty(exports, "UrlInput", { enumerable: true, get: function () { return input_1.UrlInput; } });
Object.defineProperty(exports, "PageOptionsOperation", { enumerable: true, get: function () { return input_1.PageOptionsOperation; } });
Object.defineProperty(exports, "LocalResponse", { enumerable: true, get: function () { return input_1.LocalResponse; } });
exports.internal = __importStar(require("./internal"));
exports.imageOperations = __importStar(require("./imageOperations"));
