// mindee/src/product/fr/energyBill/energyBillV1EnergySupplier.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _EnergyBillV1EnergySupplier_instances, _EnergyBillV1EnergySupplier_printableValues;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnergyBillV1EnergySupplier = void 0;
const geometry_1 = require("../../../geometry");
/**
 * The company that supplies the energy.
 */
class EnergyBillV1EnergySupplier {
    constructor({ prediction = {} }) {
        _EnergyBillV1EnergySupplier_instances.add(this);
        /** Confidence score */
        this.confidence = 0.0;
        /**
         * Contains the relative vertices coordinates (points) of a polygon containing
         * the field in the document.
         */
        this.polygon = new geometry_1.Polygon();
        this.address = prediction["address"];
        this.name = prediction["name"];
        this.pageId = prediction["page_id"];
        this.confidence = prediction["confidence"] ? prediction.confidence : 0.0;
        if (prediction["polygon"]) {
            this.polygon = prediction.polygon;
        }
    }
    /**
     * Default string representation.
     */
    toString() {
        const printable = __classPrivateFieldGet(this, _EnergyBillV1EnergySupplier_instances, "m", _EnergyBillV1EnergySupplier_printableValues).call(this);
        return ("Address: " +
            printable.address +
            ", Name: " +
            printable.name);
    }
    /**
     * Output in a format suitable for inclusion in a field list.
     */
    toFieldList() {
        const printable = __classPrivateFieldGet(this, _EnergyBillV1EnergySupplier_instances, "m", _EnergyBillV1EnergySupplier_printableValues).call(this);
        return `
  :Address: ${printable.address}
  :Name: ${printable.name}`.trimEnd();
    }
}
exports.EnergyBillV1EnergySupplier = EnergyBillV1EnergySupplier;
_EnergyBillV1EnergySupplier_instances = new WeakSet(), _EnergyBillV1EnergySupplier_printableValues = function _EnergyBillV1EnergySupplier_printableValues() {
    return {
        address: this.address ?? "",
        name: this.name ?? "",
    };
};
