// mindee/src/errors/mindeeError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MindeeHttpErrorV2 = exports.MindeeApiV2Error = exports.MindeePdfError = exports.MindeeImageError = exports.MindeeMimeTypeError = exports.MindeeError = void 0;
/**
 * Main Mindee Error custom class.
 */
class MindeeError extends Error {
    constructor(message) {
        super(message);
        this.name = "MindeeError";
    }
}
exports.MindeeError = MindeeError;
/**
 * Custom Mindee error relating to improper mimetypes in inputs.
 */
class MindeeMimeTypeError extends MindeeError {
    constructor(message) {
        super(message);
        this.name = "MindeeMimeTypeError";
    }
}
exports.MindeeMimeTypeError = MindeeMimeTypeError;
class MindeeImageError extends MindeeError {
    constructor(message) {
        super(message);
        this.name = "MindeeImageError";
    }
}
exports.MindeeImageError = MindeeImageError;
class MindeePdfError extends MindeeError {
    constructor(message) {
        super(message);
        this.name = "MindeePdfError";
    }
}
exports.MindeePdfError = MindeePdfError;
class MindeeApiV2Error extends MindeeError {
    constructor(message) {
        super(message);
        this.name = "MindeeApiV2Error";
    }
}
exports.MindeeApiV2Error = MindeeApiV2Error;
class MindeeHttpErrorV2 extends MindeeError {
    constructor(error) {
        super(`HTTP ${error.status} - ${error.title} :: ${error.code} - ${error.detail}`);
        this.status = error.status;
        this.detail = error.detail;
        this.title = error.title;
        this.code = error.code;
        this.errors = error.errors;
        this.name = "MindeeHttpErrorV2";
    }
}
exports.MindeeHttpErrorV2 = MindeeHttpErrorV2;
