// mindee/src/product/fr/idCard/internal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IdCardV2Page = exports.IdCardV2Document = exports.IdCardV2 = exports.IdCardV1Page = exports.IdCardV1Document = exports.IdCardV1 = void 0;
var idCardV1_1 = require("./idCardV1");
Object.defineProperty(exports, "IdCardV1", { enumerable: true, get: function () { return idCardV1_1.IdCardV1; } });
var idCardV1Document_1 = require("./idCardV1Document");
Object.defineProperty(exports, "IdCardV1Document", { enumerable: true, get: function () { return idCardV1Document_1.IdCardV1Document; } });
var idCardV1Page_1 = require("./idCardV1Page");
Object.defineProperty(exports, "IdCardV1Page", { enumerable: true, get: function () { return idCardV1Page_1.IdCardV1Page; } });
var idCardV2_1 = require("./idCardV2");
Object.defineProperty(exports, "IdCardV2", { enumerable: true, get: function () { return idCardV2_1.IdCardV2; } });
var idCardV2Document_1 = require("./idCardV2Document");
Object.defineProperty(exports, "IdCardV2Document", { enumerable: true, get: function () { return idCardV2Document_1.IdCardV2Document; } });
var idCardV2Page_1 = require("./idCardV2Page");
Object.defineProperty(exports, "IdCardV2Page", { enumerable: true, get: function () { return idCardV2Page_1.IdCardV2Page; } });
