// mindee/src/parsing/v2/inference.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Inference = void 0;
const inferenceModel_1 = require("./inferenceModel");
const inferenceResult_1 = require("./inferenceResult");
const inferenceFile_1 = require("./inferenceFile");
const inferenceActiveOptions_1 = require("./inferenceActiveOptions");
class Inference {
    constructor(serverResponse) {
        this.model = new inferenceModel_1.InferenceModel(serverResponse["model"]);
        this.file = new inferenceFile_1.InferenceFile(serverResponse["file"]);
        this.result = new inferenceResult_1.InferenceResult(serverResponse["result"]);
        this.activeOptions = new inferenceActiveOptions_1.InferenceActiveOptions(serverResponse["active_options"]);
    }
    toString() {
        return ("Inference\n" +
            "#########\n" +
            this.model.toString() + "\n" +
            this.file.toString() + "\n" +
            this.activeOptions.toString() + "\n" +
            this.result + "\n");
    }
}
exports.Inference = Inference;
