// mindee/src/parsing/v2/inferenceActiveOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InferenceActiveOptions = void 0;
const dataSchemaActiveOption_1 = require("./dataSchemaActiveOption");
class InferenceActiveOptions {
    constructor(serverResponse) {
        this.rag = serverResponse["rag"];
        this.rawText = serverResponse["raw_text"];
        this.polygon = serverResponse["polygon"];
        this.confidence = serverResponse["confidence"];
        this.textContext = serverResponse["text_context"];
        this.dataSchema = new dataSchemaActiveOption_1.DataSchemaActiveOption(serverResponse["data_schema"]);
    }
    toString() {
        return "Active Options\n" +
            "==============\n" +
            `:Raw Text: ${this.rawText ? "True" : "False"}\n` +
            `:Polygon: ${this.polygon ? "True" : "False"}\n` +
            `:Confidence: ${this.confidence ? "True" : "False"}\n` +
            `:RAG: ${this.rag ? "True" : "False"}\n` +
            `:Text Context: ${this.textContext ? "True" : "False"}\n\n` +
            `${this.dataSchema}\n`;
    }
}
exports.InferenceActiveOptions = InferenceActiveOptions;
