// mindee/src/parsing/v2/job.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Job = void 0;
const errorResponse_1 = require("./errorResponse");
const common_1 = require("../common");
/**
 * Job information for a V2 polling attempt.
 */
class Job {
    constructor(serverResponse) {
        this.id = serverResponse["id"];
        if (serverResponse["status"] !== undefined) {
            this.status = serverResponse["status"];
        }
        if (serverResponse["error"] !== undefined &&
            serverResponse["error"] !== null &&
            Object.keys(serverResponse["error"]).length > 0) {
            this.error = new errorResponse_1.ErrorResponse(serverResponse["error"]);
        }
        this.createdAt = (0, common_1.parseDate)(serverResponse["created_at"]);
        this.modelId = serverResponse["model_id"];
        this.pollingUrl = serverResponse["polling_url"];
        this.filename = serverResponse["filename"];
        this.resultUrl = serverResponse["result_url"];
        this.alias = serverResponse["alias"];
        this.webhooks = serverResponse["webhooks"];
    }
}
exports.Job = Job;
