// mindee/src/product/barcodeReader/barcodeReaderV1Document.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BarcodeReaderV1Document = void 0;
const common_1 = require("../../parsing/common");
const standard_1 = require("../../parsing/standard");
/**
 * Barcode Reader API version 1.0 document data.
 */
class BarcodeReaderV1Document {
    constructor(rawPrediction, pageId) {
        /** List of decoded 1D barcodes. */
        this.codes1D = [];
        /** List of decoded 2D barcodes. */
        this.codes2D = [];
        rawPrediction["codes_1d"] &&
            rawPrediction["codes_1d"].map((itemPrediction) => this.codes1D.push(new standard_1.StringField({
                prediction: itemPrediction,
                pageId: pageId,
            })));
        rawPrediction["codes_2d"] &&
            rawPrediction["codes_2d"].map((itemPrediction) => this.codes2D.push(new standard_1.StringField({
                prediction: itemPrediction,
                pageId: pageId,
            })));
    }
    /**
     * Default string representation.
     */
    toString() {
        const codes1D = this.codes1D.join("\n              ");
        const codes2D = this.codes2D.join("\n              ");
        const outStr = `:Barcodes 1D: ${codes1D}
:Barcodes 2D: ${codes2D}`.trimEnd();
        return (0, common_1.cleanOutString)(outStr);
    }
}
exports.BarcodeReaderV1Document = BarcodeReaderV1Document;
