// mindee/src/product/billOfLading/billOfLadingV1Carrier.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _BillOfLadingV1Carrier_instances, _BillOfLadingV1Carrier_printableValues;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BillOfLadingV1Carrier = void 0;
const geometry_1 = require("../../geometry");
/**
 * The shipping company responsible for transporting the goods.
 */
class BillOfLadingV1Carrier {
    constructor({ prediction = {} }) {
        _BillOfLadingV1Carrier_instances.add(this);
        /** Confidence score */
        this.confidence = 0.0;
        /**
         * Contains the relative vertices coordinates (points) of a polygon containing
         * the field in the document.
         */
        this.polygon = new geometry_1.Polygon();
        this.name = prediction["name"];
        this.professionalNumber = prediction["professional_number"];
        this.scac = prediction["scac"];
        this.pageId = prediction["page_id"];
        this.confidence = prediction["confidence"] ? prediction.confidence : 0.0;
        if (prediction["polygon"]) {
            this.polygon = prediction.polygon;
        }
    }
    /**
     * Default string representation.
     */
    toString() {
        const printable = __classPrivateFieldGet(this, _BillOfLadingV1Carrier_instances, "m", _BillOfLadingV1Carrier_printableValues).call(this);
        return ("Name: " +
            printable.name +
            ", Professional Number: " +
            printable.professionalNumber +
            ", SCAC: " +
            printable.scac);
    }
    /**
     * Output in a format suitable for inclusion in a field list.
     */
    toFieldList() {
        const printable = __classPrivateFieldGet(this, _BillOfLadingV1Carrier_instances, "m", _BillOfLadingV1Carrier_printableValues).call(this);
        return `
  :Name: ${printable.name}
  :Professional Number: ${printable.professionalNumber}
  :SCAC: ${printable.scac}`.trimEnd();
    }
}
exports.BillOfLadingV1Carrier = BillOfLadingV1Carrier;
_BillOfLadingV1Carrier_instances = new WeakSet(), _BillOfLadingV1Carrier_printableValues = function _BillOfLadingV1Carrier_printableValues() {
    return {
        name: this.name ?? "",
        professionalNumber: this.professionalNumber ?? "",
        scac: this.scac ?? "",
    };
};
