// mindee/src/parsing/standard/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PositionField = exports.PaymentDetailsField = exports.StringField = exports.TaxField = exports.Taxes = exports.Field = exports.LocaleField = exports.DateField = exports.CompanyRegistrationField = exports.ClassificationField = exports.BooleanField = exports.BaseField = exports.AmountField = exports.AddressField = void 0;
var addressField_1 = require("./addressField");
Object.defineProperty(exports, "AddressField", { enumerable: true, get: function () { return addressField_1.AddressField; } });
var amount_1 = require("./amount");
Object.defineProperty(exports, "AmountField", { enumerable: true, get: function () { return amount_1.AmountField; } });
var base_1 = require("./base");
Object.defineProperty(exports, "BaseField", { enumerable: true, get: function () { return base_1.BaseField; } });
var boolean_1 = require("./boolean");
Object.defineProperty(exports, "BooleanField", { enumerable: true, get: function () { return boolean_1.BooleanField; } });
var classification_1 = require("./classification");
Object.defineProperty(exports, "ClassificationField", { enumerable: true, get: function () { return classification_1.ClassificationField; } });
var companyRegistration_1 = require("./companyRegistration");
Object.defineProperty(exports, "CompanyRegistrationField", { enumerable: true, get: function () { return companyRegistration_1.CompanyRegistrationField; } });
var date_1 = require("./date");
Object.defineProperty(exports, "DateField", { enumerable: true, get: function () { return date_1.DateField; } });
var locale_1 = require("./locale");
Object.defineProperty(exports, "LocaleField", { enumerable: true, get: function () { return locale_1.LocaleField; } });
var field_1 = require("./field");
Object.defineProperty(exports, "Field", { enumerable: true, get: function () { return field_1.Field; } });
var tax_1 = require("./tax");
Object.defineProperty(exports, "Taxes", { enumerable: true, get: function () { return tax_1.Taxes; } });
Object.defineProperty(exports, "TaxField", { enumerable: true, get: function () { return tax_1.TaxField; } });
var text_1 = require("./text");
Object.defineProperty(exports, "StringField", { enumerable: true, get: function () { return text_1.StringField; } });
var paymentDetails_1 = require("./paymentDetails");
Object.defineProperty(exports, "PaymentDetailsField", { enumerable: true, get: function () { return paymentDetails_1.PaymentDetailsField; } });
var position_1 = require("./position");
Object.defineProperty(exports, "PositionField", { enumerable: true, get: function () { return position_1.PositionField; } });
