// mindee/src/parsing/standard/paymentDetails.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _a, _PaymentDetailsField_isKeySet;
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentDetailsField = void 0;
const field_1 = require("./field");
/**
 * Information on a single payment.
 */
class PaymentDetailsField extends field_1.Field {
    /**
     * @param {PaymentDetailsConstructor} constructor Constructor parameters.
     */
    constructor({ prediction = {}, valueKey = "iban", accountNumberKey = "account_number", ibanKey = "iban", routingNumberKey = "routing_number", swiftKey = "swift", reconstructed = false, pageId = undefined, }) {
        super({ prediction, valueKey, reconstructed, pageId });
        this.accountNumber = undefined;
        this.iban = undefined;
        this.routingNumber = undefined;
        this.swift = undefined;
        if (__classPrivateFieldGet(_a, _a, "m", _PaymentDetailsField_isKeySet).call(_a, prediction[accountNumberKey])) {
            this.accountNumber = prediction[accountNumberKey];
        }
        if (__classPrivateFieldGet(_a, _a, "m", _PaymentDetailsField_isKeySet).call(_a, prediction[ibanKey])) {
            this.iban = prediction[ibanKey];
        }
        if (__classPrivateFieldGet(_a, _a, "m", _PaymentDetailsField_isKeySet).call(_a, prediction[routingNumberKey])) {
            this.routingNumber = prediction[routingNumberKey];
        }
        if (__classPrivateFieldGet(_a, _a, "m", _PaymentDetailsField_isKeySet).call(_a, prediction[swiftKey])) {
            this.swift = prediction[swiftKey];
        }
    }
    /**
     * Default string representation.
     */
    toString() {
        let str = "";
        if (this.accountNumber !== undefined) {
            str += `${this.accountNumber}; `;
        }
        if (this.iban !== undefined) {
            str += `${this.iban}; `;
        }
        if (this.routingNumber !== undefined) {
            str += `${this.routingNumber}; `;
        }
        if (this.swift !== undefined) {
            str += `${this.swift}; `;
        }
        return str;
    }
}
exports.PaymentDetailsField = PaymentDetailsField;
_a = PaymentDetailsField, _PaymentDetailsField_isKeySet = function _PaymentDetailsField_isKeySet(value) {
    return typeof value === "string" && value !== "N/A";
};
