// mindee/src/http/apiSettingsV2.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiSettingsV2 = exports.API_V2_HOST_ENVVAR_NAME = exports.API_V2_KEY_ENVVAR_NAME = void 0;
/* eslint-disable @typescript-eslint/naming-convention */
const logger_1 = require("../logger");
const baseSettings_1 = require("./baseSettings");
const mindeeError_1 = require("../errors/mindeeError");
exports.API_V2_KEY_ENVVAR_NAME = "MINDEE_V2_API_KEY";
exports.API_V2_HOST_ENVVAR_NAME = "MINDEE_V2_API_HOST";
const DEFAULT_MINDEE_API_HOST = "api-v2.mindee.net";
class ApiSettingsV2 extends baseSettings_1.BaseSettings {
    constructor({ apiKey = "", }) {
        super();
        if (!apiKey || apiKey.length === 0) {
            this.apiKey = this.apiKeyFromEnv();
        }
        else {
            this.apiKey = apiKey;
        }
        if (!this.apiKey || this.apiKey.length === 0) {
            throw new mindeeError_1.MindeeApiV2Error("Your API V2 key could not be set, check your Client Configuration\n."
                + `You can set this using the ${exports.API_V2_KEY_ENVVAR_NAME} environment variable.`);
        }
        this.baseHeaders = {
            "User-Agent": this.getUserAgent(),
            Authorization: `${apiKey}`,
        };
    }
    apiKeyFromEnv() {
        const envVarValue = process.env[exports.API_V2_KEY_ENVVAR_NAME];
        if (envVarValue) {
            logger_1.logger.debug(`Set API key from environment: ${exports.API_V2_KEY_ENVVAR_NAME}`);
            return envVarValue;
        }
        return "";
    }
    hostnameFromEnv() {
        const envVarValue = process.env[exports.API_V2_HOST_ENVVAR_NAME];
        if (envVarValue) {
            logger_1.logger.debug(`Set the API hostname to ${envVarValue}`);
            return envVarValue;
        }
        return DEFAULT_MINDEE_API_HOST;
    }
}
exports.ApiSettingsV2 = ApiSettingsV2;
