// mindee/src/parsing/common/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseDate = exports.cleanSpecialChars = exports.floatToString = exports.extras = exports.lineSeparator = exports.cleanOutString = exports.Page = exports.Prediction = exports.PredictResponse = exports.AsyncPredictResponse = exports.OrientationField = exports.FeedbackResponse = exports.Inference = exports.ExecutionPriority = exports.ExecutionFile = exports.Execution = exports.Document = void 0;
var document_1 = require("./document");
Object.defineProperty(exports, "Document", { enumerable: true, get: function () { return document_1.Document; } });
var execution_1 = require("./execution");
Object.defineProperty(exports, "Execution", { enumerable: true, get: function () { return execution_1.Execution; } });
var executionFile_1 = require("./executionFile");
Object.defineProperty(exports, "ExecutionFile", { enumerable: true, get: function () { return executionFile_1.ExecutionFile; } });
var executionPriority_1 = require("./executionPriority");
Object.defineProperty(exports, "ExecutionPriority", { enumerable: true, get: function () { return executionPriority_1.ExecutionPriority; } });
var inference_1 = require("./inference");
Object.defineProperty(exports, "Inference", { enumerable: true, get: function () { return inference_1.Inference; } });
var feedbackResponse_1 = require("./feedback/feedbackResponse");
Object.defineProperty(exports, "FeedbackResponse", { enumerable: true, get: function () { return feedbackResponse_1.FeedbackResponse; } });
var orientation_1 = require("./orientation");
Object.defineProperty(exports, "OrientationField", { enumerable: true, get: function () { return orientation_1.OrientationField; } });
var asyncPredictResponse_1 = require("./asyncPredictResponse");
Object.defineProperty(exports, "AsyncPredictResponse", { enumerable: true, get: function () { return asyncPredictResponse_1.AsyncPredictResponse; } });
var predictResponse_1 = require("./predictResponse");
Object.defineProperty(exports, "PredictResponse", { enumerable: true, get: function () { return predictResponse_1.PredictResponse; } });
var prediction_1 = require("./prediction");
Object.defineProperty(exports, "Prediction", { enumerable: true, get: function () { return prediction_1.Prediction; } });
var page_1 = require("./page");
Object.defineProperty(exports, "Page", { enumerable: true, get: function () { return page_1.Page; } });
var summaryHelper_1 = require("./summaryHelper");
Object.defineProperty(exports, "cleanOutString", { enumerable: true, get: function () { return summaryHelper_1.cleanOutString; } });
Object.defineProperty(exports, "lineSeparator", { enumerable: true, get: function () { return summaryHelper_1.lineSeparator; } });
exports.extras = __importStar(require("./extras"));
var summaryHelper_2 = require("./summaryHelper");
Object.defineProperty(exports, "floatToString", { enumerable: true, get: function () { return summaryHelper_2.floatToString; } });
Object.defineProperty(exports, "cleanSpecialChars", { enumerable: true, get: function () { return summaryHelper_2.cleanSpecialChars; } });
var dateParser_1 = require("./dateParser");
Object.defineProperty(exports, "parseDate", { enumerable: true, get: function () { return dateParser_1.parseDate; } });
