// mindee/src/product/internal.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.us = exports.resume = exports.receipt = exports.passport = exports.nutritionFactsLabel = exports.multiReceiptsDetector = exports.invoiceSplitter = exports.invoice = exports.internationalId = exports.ind = exports.generated = exports.fr = exports.financialDocument = exports.driverLicense = exports.deliveryNote = exports.custom = exports.cropper = exports.businessCard = exports.billOfLading = exports.barcodeReader = void 0;
exports.barcodeReader = __importStar(require("./barcodeReader/internal"));
exports.billOfLading = __importStar(require("./billOfLading/internal"));
exports.businessCard = __importStar(require("./businessCard/internal"));
exports.cropper = __importStar(require("./cropper/internal"));
exports.custom = __importStar(require("./custom/internal"));
exports.deliveryNote = __importStar(require("./deliveryNote/internal"));
exports.driverLicense = __importStar(require("./driverLicense/internal"));
exports.financialDocument = __importStar(require("./financialDocument/internal"));
exports.fr = __importStar(require("./fr/internal"));
exports.generated = __importStar(require("./generated/internal"));
exports.ind = __importStar(require("./ind/internal"));
exports.internationalId = __importStar(require("./internationalId/internal"));
exports.invoice = __importStar(require("./invoice/internal"));
exports.invoiceSplitter = __importStar(require("./invoiceSplitter/internal"));
exports.multiReceiptsDetector = __importStar(require("./multiReceiptsDetector/internal"));
exports.nutritionFactsLabel = __importStar(require("./nutritionFactsLabel/internal"));
exports.passport = __importStar(require("./passport/internal"));
exports.receipt = __importStar(require("./receipt/internal"));
exports.resume = __importStar(require("./resume/internal"));
exports.us = __importStar(require("./us/internal"));
